### PR TITLE
Fix certificate request logic

### PR DESCRIPTION
**Pull request by NobleKangaroo**
### Preface
Using the Let's Encrypt addon, a new certificate cannot be requested (e.g. if you add or remove domains) if you've already generated a certificate. This is problematic if you have renamed your host, or moved to a new domain, or in the case of any of a number of other reasons that you need to generate a new certificate.

### Original configuration and initial request result
Below is a redacted copy of my configuration, likely similar to most users:
```yaml
email: sysadmin@example.com
domains:
  - hass.example.com
certfile: fullchain.pem
keyfile: privkey.pem
challenge: dns
dns:
  provider: dns-cloudflare
  cloudflare_email: admin@example.com
  cloudflare_api_key: 0123456789abcdef0123456789abcdef01234
```

Resultant log data following starting the Let's Encrypt addon:
```log
[17:31:59] INFO: Selected DNS Provider: dns-cloudflare
[17:31:59] INFO: Use propagation seconds: 60
[17:32:03] WARNING: Use CloudFlare global key (not recommended!)
Saving debug log to /var/log/letsencrypt/letsencrypt.log
Plugins selected: Authenticator dns-cloudflare, Installer None
Obtaining a new certificate
Performing the following challenges:
dns-01 challenge for hass.example.com
Waiting 60 seconds for DNS changes to propagate
Waiting for verification...
Cleaning up challenges
Non-standard path(s), might not work with crontab installed by your operating system package manager
IMPORTANT NOTES:
 - Congratulations! Your certificate and chain have been saved at:
   /data/letsencrypt/live/hass.example.com/fullchain.pem
   Your key file has been saved at:
   /data/letsencrypt/live/hass.example.com/privkey.pem
   Your cert will expire on 2020-05-28. To obtain a new or tweaked
   version of this certificate in the future, simply run certbot
   again. To non-interactively renew *all* of your certificates, run
   "certbot renew"
 - Your account credentials have been saved in your Certbot
   configuration directory at /data/letsencrypt. You should make a
   secure backup of this folder now. This configuration directory will
   also contain certificates and private keys obtained by Certbot so
   making regular backups of this folder is ideal.
 - If you like Certbot, please consider supporting our work by:
   Donating to ISRG / Let's Encrypt:   https://letsencrypt.org/donate
   Donating to EFF:                    https://eff.org/donate-le
```

### Modified configuration and subsequent request result
If we change the configuration to use a different domain, such as moving to a subdomain, using a different domain, or adding additional domains, the addon fails to request a new certificate (for reasons later discussed).
```yaml
email: sysadmin@example.com
domains:
  - hass.example.com
  - homeassistant.example.com    # Note the additional second domain here
certfile: fullchain.pem
keyfile: privkey.pem
challenge: dns
dns:
  provider: dns-cloudflare
  cloudflare_email: admin@example.com
  cloudflare_api_key: 0123456789abcdef0123456789abcdef01234
```

Resultant log data following starting the Let's Encrypt addon:
```log
[17:37:53] INFO: Selected DNS Provider: dns-cloudflare
[17:37:53] INFO: Use propagation seconds: 60
[17:37:57] WARNING: Use CloudFlare global key (not recommended!)
Saving debug log to /var/log/letsencrypt/letsencrypt.log
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Processing /data/letsencrypt/renewal/hass.example.com.conf
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Cert not yet due for renewal
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
The following certs are not due for renewal yet:
  /data/letsencrypt/live/hass.example.com/fullchain.pem expires on 2020-05-28 (skipped)
No renewals were attempted.
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
```

### Investigation of the issue and proposed resolution
The issue appears to be caused by the addon's `run.sh` script's simplistic check for whether or not a certificate needs to be requested.
```sh
105| ### /mnt/data/supervisor/addons/core/letsencrypt/data/run.sh
106| # Generate new certs
107| if [ ! -d "$CERT_DIR/live" ]; then
108|     DOMAIN_ARR=()
109|     for line in $DOMAINS; do
110|         DOMAIN_ARR+=(-d "$line")
111|     done
112| 
113|     echo "$DOMAINS" > /data/domains.gen
114|     if [ "$CHALLENGE" == "dns" ]; then
115|         certbot certonly --non-interactive --config-dir "$CERT_DIR" --work-dir "$WORK_DIR" "${PROVIDER_ARGUMENTS[@]}" --email "$EMAIL" --agree-tos --config-dir "$CE
116|     else
117|         certbot certonly --non-interactive --standalone --email "$EMAIL" --agree-tos --config-dir "$CERT_DIR" --work-dir "$WORK_DIR" --preferred-challenges "$CHALLE
118|     fi
119| else
120|     certbot renew --non-interactive --config-dir "$CERT_DIR" --work-dir "$WORK_DIR" --preferred-challenges "$CHALLENGE"
121| fi
122| 
123| # copy certs to store
124| cp "$CERT_DIR"/live/*/privkey.pem "/ssl/$KEYFILE"
125| cp "$CERT_DIR"/live/*/fullchain.pem "/ssl/$CERTFILE"
```

Lines 108 through 118, where the requesting of a new certificate takes place, are ignored if the `live` directory exists (generated when the first certificate is issued). I understand this is to to avoid re-requesting a certificate every time the addon is started, however through passing some flags to the certbot command the same result can be achieved while also allowing for a certificate to be expanded.

I recommend changing the code to always use `certbot certonly` instead of `certbot renew` as the latter does not provide a facility to expand certificates. Through passing command flags to `certbot certonly`, you can still ensure a new certificate is only requested when the previous certificate is close to expiration (30 days) but also request a new one if domains are changed. This PR will suffice to accomplish that through the use of the `--keep-until-expiring` flag (to ensure that we don't request a certificate too early) and the `--expand` flag (to ensure that domains being changed results in a new certificate being issued). See the code for the changes made, which also includes minor cleanups to improve readability of the commands and duplicate passing of the `--config-dir` and `--work-dir` flags (in the `if [ "$CHALLENGE" == "dns" ]; then` section).

I validated the changes this pull request implements by using a modified `run.sh` and starting the addon.

### Modified configuration (and run.sh) and subsequent request result
```yaml
email: sysadmin@example.com
domains:
  - hass.example.com
  - homeassistant.example.com    
  - home-assistant.example.com    # Note the additional third domain here
certfile: fullchain.pem
keyfile: privkey.pem
challenge: dns
dns:
  provider: dns-cloudflare
  cloudflare_email: admin@example.com
  cloudflare_api_key: 0123456789abcdef0123456789abcdef01234
```

Resultant log data following starting the Let's Encrypt addon:
Shown below, the script now properly captures the two (one from the failed attempt, and one from from the second edited config above) domains and adds them to the new certificate request.
```log
[19:12:35] INFO: Selected DNS Provider: dns-cloudflare
[19:12:35] INFO: Use propagation seconds: 60
[19:12:39] WARNING: Use CloudFlare global key (not recommended!)
Saving debug log to /var/log/letsencrypt/letsencrypt.log
Plugins selected: Authenticator dns-cloudflare, Installer None
Renewing an existing certificate
Performing the following challenges:
dns-01 challenge for homeassistant.example.com
dns-01 challenge for home-assistant.example.com
Waiting 60 seconds for DNS changes to propagate
Waiting for verification...
Cleaning up challenges
IMPORTANT NOTES:
 - Congratulations! Your certificate and chain have been saved at:
   /data/letsencrypt/live/hass.example.com/fullchain.pem
   Your key file has been saved at:
   /data/letsencrypt/live/hass.example.com/privkey.pem
   Your cert will expire on 2020-05-28. To obtain a new or tweaked
   version of this certificate in the future, simply run certbot
   again. To non-interactively renew *all* of your certificates, run
   "certbot renew"
 - If you like Certbot, please consider supporting our work by:
   Donating to ISRG / Let's Encrypt:   https://letsencrypt.org/donate
   Donating to EFF:                    https://eff.org/donate-le
```

### Same configuration and final request result
For the final test, I left the last edited configuration in place.
```yaml
email: sysadmin@example.com
domains:
  - hass.example.com
  - homeassistant.example.com    
  - home-assistant.example.com
certfile: fullchain.pem
keyfile: privkey.pem
challenge: dns
dns:
  provider: dns-cloudflare
  cloudflare_email: admin@example.com
  cloudflare_api_key: 0123456789abcdef0123456789abcdef01234
```

I started the addon again and was presented with a proper exit, without requesting a new certificate.
```log
[19:19:55] INFO: Selected DNS Provider: dns-cloudflare
[19:19:55] INFO: Use propagation seconds: 60
[19:19:59] WARNING: Use CloudFlare global key (not recommended!)
Saving debug log to /var/log/letsencrypt/letsencrypt.log
Plugins selected: Authenticator dns-cloudflare, Installer None
Cert not yet due for renewal
Keeping the existing certificate
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Certificate not yet due for renewal; no action taken.
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
```

### CODE REVIEW

File: letsencrypt/data/run.sh [hunk 1]
```diff
@@ -123,6 +123,7 @@ else
         --preferred-challenges "$CHALLENGE" "${DOMAIN_ARR[@]}" --standalone
 fi
 
-# copy certs to store
-cp "$CERT_DIR"/live/*/privkey.pem "/ssl/$KEYFILE"
-cp "$CERT_DIR"/live/*/fullchain.pem "/ssl/$CERTFILE"
+# Get the last modified cert directory and copy the cert and private key to store
+CERT_DIR=$(ls -td $CERT_DIR/live/* | head -1)
```
Comment (NobleKangaroo) [inline letsencrypt/data/run.sh]: In the event there's a directory for a certificate received for a given domain (e.g. hass.example.com) and the configuration is updated to add additional certificates to that domain (e.g. homeassistant.example.com), Let's Encrypt appends `-0001`, `-0002`, etc to the end of the directory to ensure previously issued certificates are not overwritten. We simply need to make sure we're using the the most recently modified directory, as it'll be the one certbot just put the files in.

File: letsencrypt/data/run.sh [hunk 2]
```diff
@@ -123,6 +123,7 @@ else
...
+cp "$CERT_DIR"/privkey.pem "/ssl/$KEYFILE"
+cp "$CERT_DIR"/fullchain.pem "/ssl/$CERTFILE"
```
Comment (NobleKangaroo) [inline letsencrypt/data/run.sh]: Removed the `live/*/` bit as the code above figures out which one to use.

File: letsencrypt/data/run.sh [hunk 3]
```diff
@@ -124,6 +124,6 @@ else
 fi
 
 # Get the last modified cert directory and copy the cert and private key to store
-CERT_DIR=$(ls -td $CERT_DIR/live/* | head -1)
+CERT_DIR=$(realpath(ls -td $CERT_DIR/live/*/ | head -1))
```
Comment (NobleKangaroo) [inline letsencrypt/data/run.sh]: (outside of the typo here) README file was being returned, assume it gets laid down after the cert files. Using trailing slash to only look for directories, and `realpath` to remove the trailing slash.

Code check fails because of `ls` call (really.. it's too damn much to try to sort directories by modification date using the `find` command) and not-quoted parameter (we can fix that if needed, no big deal).

`CERT_DIR=$(realpath "$(ls -td $CERT_DIR/live/*/ | head -1)")` would be the quoted variant.

Comment (pvizeli) [reply letsencrypt/data/run.sh]: > CERT_DIR=$(realpath "$(ls -td $CERT_DIR/live/*/ | head -1)") would be the quoted variant.

let do this, so we can merge if he pass the CI.

File: letsencrypt/data/run.sh [hunk 4]
```diff
@@ -124,6 +124,6 @@ else
 fi
 
 # Get the last modified cert directory and copy the cert and private key to store
-CERT_DIR=$(find $CERT_DIR/live/* -type d -maxdepth 1 | awk '{ "stat -c '%Y' "$0 | getline ts; if (ts >= ots) printf "%d\t%s\n", ts, $0; ots=ts; }' | sort -k1 | tail -1 | awk '{ print $2 }')
+CERT_DIR=$(find $CERT_DIR/live/* -type d -maxdepth 1 | awk '{ "stat -c '%Y' "$0 | getline ts; if (ts >= ots) printf "%d\t%s\n", ts, $0; ots=ts; }' | sort -nk1 | tail -1 | awk '{ print $2 }')
```
Comment (NobleKangaroo) [inline letsencrypt/data/run.sh]: It's not the prettiest (and personally it's a bit overkill, just to be 100% on ShellCheck) but it works. 😄

`find $CERT_DIR/live/* -type d -maxdepth 1 `
Find all directories in /data/letsencrypt/live/

`awk '{ "stat -c '%Y' "$0 | getline ts; if (ts >= ots) printf "%d\t%s\n", ts, $0; ots=ts; }' `
Get the second since Epoch when the directory was modified and print it in the format "seconds_since_epoch(tab)directory_path"

`sort -nk1 | tail -1 | awk '{ print $2 }'`
Sort them numerically, by the first field (the Epoch), show only the last line, and show only the second field (the full path)

While I found that it worked in my own environment, it may be a good idea to tinker with it and test that there aren't any obvious edge cases that I didn't come across.

File: letsencrypt/data/run.sh [hunk 5]
```diff
@@ -124,6 +124,6 @@ else
 fi
 
 # Get the last modified cert directory and copy the cert and private key to store
-CERT_DIR=$(find $CERT_DIR/live/* -type d -maxdepth 1 | awk '{ "stat -c '%Y' "$0 | getline ts; if (ts >= ots) printf "%d\t%s\n", ts, $0; ots=ts; }' | sort -nk1 | tail -1 | awk '{ print $2 }')
+CERT_DIR=$(find $CERT_DIR/live/* -type d -maxdepth 1 | awk '{ "stat -c '%Y' "$0 | getline ts; printf "%d\t%s\n", ts, $0 }' | sort -nk1 | tail -1 | awk '{ print $2 }')
```
Comment (NobleKangaroo) [inline letsencrypt/data/run.sh]: No need to do an `if..then` comparison in awk as we're following up with a sort directly after.

**Comment by NobleKangaroo** [comment]
I *think* that covers my OCD tinglings. 😆 Feel free to suggest edits.

**Comment by pvizeli** [comment]
I like to keep it simple :)

**Comment by pvizeli** [comment]
This should work. What do you think?

**Comment by NobleKangaroo** [comment]
> This should work. What do you think?

Great, I was trying to avoid the code test failures going forward (since the code is now not passing) but if that works for you, that's fine by me!

How about the rest of the PR? Still need to call `certbot certonly` instead of `certbot renew` (unless that's still in the pipeline to be merged).

**Comment by pvizeli** [comment]
We can avoid a shellcheck issue if that make no sense

**Comment by pvizeli** [comment]
So this code is fine for you and reflect the issue they it should solve now?

**Comment by NobleKangaroo** [comment]
> We can avoid a shellcheck issue if that make no sense
> So this code is fine for you and reflect the issue they it should solve now?

Looks good here. Will keep an eye on it in case there's any unforeseen issues.
